pontoon/base/placeables.py

```diff
@@ -2,39 +2,56 @@
 
 
 class PythonFormatNamedString:
-    parser = r"(%\([[\w\d!.,[\]%:$<>+\-= ]*\)[+|-|0\d+|#]?[.\d+]?[s|d|e|f|g|o|x|c|%])"
+    parser = re.compile(
+        r"(%\([[\w\d!.,[\]%:$<>+\-= ]*\)[+|-|0\d+|#]?[.\d+]?[s|d|e|f|g|o|x|c|%])",
+        re.IGNORECASE,
+    )
 
 
 class PythonFormatString:
-    parser = r"(\{{?[\w\d!.,[\]%:$<>+-= ]*\}?})"
+    parser = re.compile(r"(\{{?[\w\d!.,[\]%:$<>+-= ]*\}?})")
 
 
 class PythonFormattingVariable:
-    parser = r"(%(%|(\([^)]+\)){0,1}[-+0#]{0,1}(\d+|\*){0,1}(\.(\d+|\*)){0,1}[hlL]{0,1}[diouxXeEfFgGcrs]{1}))"
+    parser = re.compile(
+        r"(%(%|(\([^)]+\)){0,1}[-+0#]{0,1}(\d+|\*){0,1}(\.(\d+|\*)){0,1}[hlL]{0,1}[diouxXeEfFgGcrs]{1}))"
+    )
     match_index = 0
 
 
 class FluentTerm:
-    parser = r"({ ?-[^}]* ?})"
+    parser = re.compile(r"({ ?-[^}]* ?})")
 
 
 class FluentParametrizedTerm:
-    parser = r"({ ?-[^}]*([^}]*: ?[^}]*) ?})"
+    parser = re.compile(r"({ ?-[^}]*([^}]*: ?[^}]*) ?})")
 
 
 class FluentFunction:
-    parser = r"({ ?[A-W0-9\-_]+[^}]* ?})"
+    parser = re.compile(r"({ ?[A-W0-9\-_]+[^}]* ?})")
+
+
+class JavaFormattingVariable:
+    parser = re.compile(
+        r"(?x){[0-9]+(,\s*(number(,\s*(integer|currency|percent|[-0#.,E;%\u2030\u00a4']+)?)?|(date|time)(,\s*(short|medium|long|full|.+?))?|choice,([^{]+({.+})?)+)?)?}"
+    )
```

Comment:
```suggestion
    parser = re.compile(
        r"""
        {
          [0-9]+
          (,\s*(number
            (,\s*(integer | currency | percent | [-0#.,E;%\u2030\u00a4']+)?)?
            | (date | time)(,\s*(short | medium | long | full | .+?))?
            | choice,([^{]+({.+})?)+
          )?)?
        }
        """,
        re.X,
    )
```

When using `re.compile()`, it's a bit cleaner to replace the inline `(?x)` with [re.X](https://docs.python.org/3/library/re.html#re.X), which allows whitespace.

The above is equivalent to the given regex, but isn't it a bit broken for date/time/choice?